pubsub/gossipsub/README.md

```diff
@@ -1,7 +1,373 @@
+and adds them to the mesh, notifying them with a control message. When
+it leaves the topic, it notifies its peers and forgets the mesh for
+the topic.
+
+The mesh is maintained with the following periodic stabilization
+algorithm:
+
+```
+at each peer:
+  loop:
+    if |peers| < D_low:
+       select D - |peers| non-mesh peers at random and add them to the mesh
+    if |peers| > D_high:
+       select |peers| - D mesh peers at random and remove them from the mesh
+    sleep t
+```
+The parameters of the algorithm are `D` which is the target degree,
+and two relaxed degree parameters `D_low` and `D_high` which represent
+admissible mesh degree bounds.
+
+### gossipsub: The gossiping mesh router
+
+The meshsub router offsers a baseline construction with good amplification
+control properties, which we augment with _gossip_ about message flow.
+The gossip is emited to random subsets of peers not in the mesh, similar
+to randomsub, and it allows us to propagate _metadata_ about message flow
+throughout the network. The metadata can be arbitrary, but as a baseline
+we include the message ids of seen messages in the last few seconds.
+The messages are cached, so that peers receiving the gossip can request
+them out of band.
+
+The router can use this metadata to improve the mesh, for instance an
+[episub](episub.md) router built on top of gossipsub can create
+epidemic broadcast trees.  Beyond that, the metadata can restart
+message transmission at different points in the overlay to rectify
+downstream message loss. Or it can simply jump hops oppurtunistically
+and accelerate message transmission for peers who are at some distance
+in the mesh.
+
+Essentially, gossipsub is a blend of meshsub for data and randomsub
+for mesh metadata. It provides bounded degree and amplification factor
+with the meshsub construction and augments it using gossip propagation
+of metadata with the randomsub technique.
+
+
+## The gossipsub protocol
+
+We can now provide a specification of the pubsub protocol by sketching
+out the router implementation.  The router is backwards compatible
+with floodsub, as it accepts floodsub peers and behaves like floodsub
+towards them.
+
+### Control messages
+
+The protocol defines four control messages:
+- `GRAFT`: graft a mesh link; notifies the peer that it has been added to the local mesh view.
+- `PRUNE`: prune a mesh link; notifies the peer that it has been removed from the local mesh view.
+- `IHAVE`: gossip; notifies the peer that the following messages were recently seen and are available on request.
+- `IWANT`: requests the out of band transmission of messages announced in an `IHAVE` message.
+
+### Router state
+
+The router maintains the following state:
+- `peers`: a set of all known peers; `peers.gossipsub` denotes the gossipsub peers
+   while `peers.floodsub` denotes the floodsub peers.
+- `mesh`: the overlay meshes as a map of topic to lists of peers.
```

Comment:
Is it correct that all set elements in `mesh` are also in `peers.gossipsub`, but not necessarily the other way around?

Reply:
yes, correct.